File: Java/백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int n, m;
    static int[][] arr;
    static int[] rowArr = {1, -1, 0, 0};
    static int[] colArr = {0, 0, 1, -1};
    static Queue<Point> queue = new LinkedList<>();

    static class Point {
        int row, col, day;
        public Point(int row, int col, int day) {
            this.row = row;
            this.col = col;
            this.day = day;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        arr = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
                if (arr[i][j] == 1) {
                    queue.add(new Point(i, j, 0));
                }
            }
        }

        bfs();
    }

    static void bfs() {
        int day = 0;

        while (!queue.isEmpty()) {
            Point p = queue.poll();
            day = p.day;

            for (int i = 0; i < 4; i++) {
                int nr = p.row + rowArr[i];
                int nc = p.col + colArr[i];

                if (0 <= nr && nr < n && 0 <= nc && nc < m) {
                    if (arr[nr][nc] == 0) {
                        arr[nr][nc] = 1;
                        queue.add(new Point(nr, nc, day + 1));
                    }
                }
            }
        }

        if (check()) {
            System.out.println(day);
        } else {
            System.out.println(-1);
        }
    }

    static boolean check() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr[i][j] == 0)
                    return false;
            }
        }
        return true;
    }
}

File: C11/백준/Gold/7576. 토마토/토마토.c
#include <stdio.h>

#define MAX 1000

typedef struct {
    int row, col, day;
} Point;

int n, m;
int arr[MAX][MAX];
int rowArr[4] = {1, -1, 0, 0};
int colArr[4] = {0, 0, 1, -1};

Point queue[MAX * MAX];
int front = 0, rear = 0;

void enqueue(Point p) {
    queue[rear++] = p;
}

Point dequeue() {
    return queue[front++];
}

int isEmpty() {
    return front == rear;
}

int check() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (arr[i][j] == 0) return 0;
        }
    }
    return 1;
}

void bfs() {
    int day = 0;

    while (!isEmpty()) {
        Point p = dequeue();
        day = p.day;

        for (int i = 0; i < 4; i++) {
            int nr = p.row + rowArr[i];
            int nc = p.col + colArr[i];

            if (nr >= 0 && nr < n && nc >= 0 && nc < m) {
                if (arr[nr][nc] == 0) {
                    arr[nr][nc] = 1;
                    enqueue((Point){nr, nc, day + 1});
                }
            }
        }
    }

    if (check()) {
        printf("%d\n", day);
    } else {
        printf("-1\n");
    }
}

int main() {
    scanf("%d %d", &m, &n);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            scanf("%d", &arr[i][j]);
            if (arr[i][j] == 1) {
                enqueue((Point){i, j, 0});
            }
        }
    }

    bfs();
    
    return 0;
}